### DISCUSSION TITLE
Replace Footnotes Header with a horizontal line

### DISCUSSION THREAD

**Discussion by shamoilarsi** [facebook/docusaurus#10374] [Q&A] [answered]
on adding footnotes to my docs website, a header `Footnotes` is added by default. 
<img width="684" alt="Screenshot 2024-08-05 at 8 51 19 PM" src="https://github.com/user-attachments/assets/bee941df-e4c5-48a2-9d77-f078e4398a0b">

I would like to have a horizontal line instead of the header as seen in Uniswap [here](https://docs.uniswap.org/concepts/uniswap-protocol)
<img width="910" alt="Screenshot 2024-08-05 at 8 56 02 PM" src="https://github.com/user-attachments/assets/c01a0169-7ca7-4860-9471-7f645412e9ea">

A possible solution is that i could add a `---` on the `.md` files and hide the header from the `custom.css` file by using 
```
.footnotes h2 {
  display: none;
}
```
but this looks like a hacky way of doing it. It is also not the way Uniswap achieved it (confirmed from their GitHub repo)

any suggestions on how I can achieve this?

**Comment by slorber** [comment]
Uniswap site is using Docusaurus v2 (using MDX v1)

Using a separator like this is how MDX v1 used to do. But now in MDX v2 (used by Docusaurus v3) it creates a heading.

I'm not sure how you can revert to the old behavior, apart writing a remark plugin that replaces that heading with a separator. Any idea @wooorm ?

Note that we expose `config.markdown.remarkRehypeOptions` that permit to control a bit this behavior: 
- https://github.com/facebook/docusaurus/issues/9502
- https://github.com/remarkjs/remark-rehype#options

Maybe try `footnoteLabelTagName: 'hr'` ?

**Reply by wooorm** [reply, marked as answer]
You can configure things. Read the readmes of https://github.com/remarkjs/remark-gfm and https://github.com/remarkjs/remark-rehype indeed. Particularly the CSS section in `remark-gfm`.

**Reply by shamoilarsi** [reply]
@slorber @wooorm hey guys thanks for the reply.

i tried `footnoteLabelTagName: 'hr'` but it doesn't seem to work
```
hr is a void element tag and must neither have `children` nor use `dangerouslySetInnerHTML`.
```

I read the readmes of the `remark-gfm` and `remark-rehype` but I couldn't find anything that would completely solve my issue. 
In the CSS sections I found a way to hide the label itself by editing the `.sr-only` class although I need to replace it with a horizontal line as well without editing the `.md`/`.mdx` file.

I've already tired the method of hiding the "Footnotes" header using CSS and adding a `---` in the `.md` file. It works although it felt a bit hacky and thought maybe there was some config option that I overlooked.

**Reply by slorber** [reply]
@shamoilarsi I just applied @wooorm suggestion and it worked immediately for me 🤷‍♂️ 


Just copy that CSS: https://github.com/micromark/micromark-extension-gfm-footnote#css

https://stackblitz.com/edit/github-wjfzlc?file=src%2Fpages%2Findex.mdx,src%2Fcss%2Fcustom.css

![CleanShot 2024-08-12 at 12 57 52@2x](https://github.com/user-attachments/assets/6a03dd29-d5da-4c4a-b94b-41265cf6657e)

The only difference it that the h2 becomes a separator, and it's not using `<hr/>` under the hood.

**Reply by shamoilarsi** [reply]
@slorber ahh you're correct, it works. 
I remember trying this other day, not sure what went wrong then. I might have been too focused to achieve it with `hr` as seen on Uniswap.

Anyway thanks a lot @slorber and @wooorm 🙏

**Reply by wooorm** [reply]
You can use CSS to create a line, too.
You do not need an `hr` element.

**Reply by shamoilarsi** [reply]
@wooorm yep, I had thought of this but felt a bit hacky which is why I was focused on making it work with `hr`, not to mention Uniswap had already achieved it with `hr` (by using the older version though). 

Now that I know the documentation suggests the CSS method too (thanks to you and slorber), maybe its not as hacky as I made it out to be. 

By the way, if I wanted to contribute (if you feel its a decent addition) by fixing this with a `hr` specifically in docusaurus (maybe MDX v2 directly?), how would you recommend going about it?

**Reply by wooorm** [reply]
I dont think there's a fix because there's no bug. Like, what do you want? And why? The markup is semantic. It's not particularly bad that there's a "footnotes" heading for footnotes.
What do you want? Why do you not want accessible text? Why do you not want to use CSS?

**Reply by shamoilarsi** [reply]
my bad, I don't mean the "footnotes" header is bad as such. I just had a requirement of having a horizontal line instead.

> What do you want?

Ideally, I would like an option to choose between a header and a line from the config.

> Why do you not want accessible text?

mainly for looks, personally for me a line is more appealing than a header and I'm sure there are others too.

> Why do you not want to use CSS?

not at all against using CSS. Just thought that having an option in the config which would add a `hr` element (like in MDX v1) would be more appropriate than the CSS equivalent.

edited the previous message for better clarity.

I'm just curious if I wanted to contribute to docusaurus by adding an option in the config for the user to change between a header and a horizontal line,

1. do you think its something that would be useful?
2. if yes, how would you recommend going about it?

**Reply by wooorm** [reply]
This has little to do with MDX. Or Docusaurus. I do not maintain Docusaurus. I have no opinion on whether options should go here.

The difference you observe, is because footnotes were added to GFM: you can use them here on GH now. And this is how they work. The tools in question, namely https://github.com/remarkjs/remark-rehype, follow how this works.
Footnotes before were without a standard.
`remark-rehype` has options, did you see them? https://github.com/remarkjs/remark-rehype#options

Then, there are also plugins. Plugins can already do many things. They can change things. You can write a plugin to change this heading to a `hr` if you want?

> would semantically be more appropriate.

I doubt this? Why? Can you reason for this?
I can understand a `hr` over *nothing*, but I cannot understand an `hr` over a heading with words?

**Reply by shamoilarsi** [reply]
> remark-rehype has options, did you see them? https://github.com/remarkjs/remark-rehype#options

yep yep I did. When i said I'd like to add that option, I was referring to these options only. 


thanks for the explanation, I'd read more about this. 

 > I doubt this? Why? Can you reason for this?

I understand drawing a line with CSS and `border-top` could be more customisable as compared to an `hr` element, although strictly semantically speaking having a `hr` tag would make more sense.
for eg - you could have a footer section wrapped in a `div` tag but wrapping it around the `footer` tag is semantically better. 


I think I understand your point. 
You're comparing a "footnotes" header (which is visually hidden) and a horizontal line using CSS with just a `hr` tag. 
While I'm comparing just a horizontal line using CSS with a `hr` element

yeah yours is closer to the implementation in this case it makes more sense 👍

**Reply by wooorm** [reply]
> You're comparing a "footnotes" header […] with just a hr tag. While I'm comparing just a horizontal line using CSS with a hr element.

Right, indeed!

`hr` is for a thematic break. A section break. And headings seem like that to me too, but better.
Importantly, `hr` *used* to be a presentational element without meaning, it was “retconned” to have a meaning.

`remark-rehype`/`mdast-util-to-hast` already has options to change that element, `footnoteLabel` for content, `footnoteLabelProperties` for props, `footnoteLabelTagName` for the tag name.
To do what you want, `footnoteLabelTagName` can be set to `'hr'`, `footnoteLabelProperties` to `{}`, then the only thing missing is `footnoteLabel` which you want empty, but there’s already the nully meaning for the default.
One idea for that is to support `''` as no content. Or we should accept children nodes, and then `[]` could be used.
If you have a strong use case for why you do not want content in this footnote label, why this `hr` is a legitimate use case, I would appreciate discussing this in https://github.com/syntax-tree/mdast-util-to-hast!
